arelle/plugin/validate/NL/rules/fr_nl.py

```diff
@@ -70,6 +118,59 @@ def rule_fr_nl_1_03(
                 )
 
 
+@validation(
+    hook=ValidationHook.XBRL_FINALLY,
+    disclosureSystems=[
+        DISCLOSURE_SYSTEM_NT16,
+        DISCLOSURE_SYSTEM_NT17,
+        DISCLOSURE_SYSTEM_NT18,
+    ],
+)
+def rule_fr_nl_1_04(
+        pluginData: PluginValidationDataExtension,
+        val: ValidateXbrl,
+        *args: Any,
+        **kwargs: Any,
+) -> Iterable[Validation] | None:
+    """
+    FR-NL-1.04: Disallowed character references MUST NOT be used. The use of character references (e.g. &#1080;)
+    is not allowed unless it concerns numeric character references within the allowed set of characters and except
+    for the entity references listed here: &lt; &gt; &amp; &apos; &quot;
+    """
+    pattern = regex.compile(r"""
+          & \#     (?<dec>[0-9]+)         ;
+        | & \#[Xx] (?<hex>[0-9A-Fa-f]+)   ;
+        | &        (?<named>[0-9A-Za-z]+) ;
+    """, regex.VERBOSE)
+    modelXbrl = val.modelXbrl
+    sourceFileLines = []
+    for doc in modelXbrl.urlDocs.values():
+        if doc.type == ModelDocument.Type.INSTANCE:
+            with modelXbrl.fileSource.file(doc.filepath)[0] as file:
+                for i, line in enumerate(file):
+                    for match in regex.finditer(pattern, line):
+                        decimalValue = None
+                        if (numericMatch := match.group('hex')) is not None:
+                            decimalValue = int(numericMatch, 16)
+                        if (numericMatch := match.group('dec')) is not None:
+                            decimalValue = int(numericMatch)
+                        if decimalValue is not None:
+                            if any(min <= decimalValue <= max
+                                   for min, max in UNICODE_CHARACTER_DECIMAL_RANGES):
+                                continue  # numeric references in certain ranges are allowed
+                        if (namedMatch := match.group('named')) is not None:
+                            if namedMatch in ALLOWED_NAMED_CHARACTER_REFS:
+                                continue  # certain named references are allowed
+                        sourceFileLines.append((doc.filepath, i + 1))
+    if len(sourceFileLines) > 0:
+        yield Validation.error(
+            codes='NL.FR-NL-1.04',
+            msg=_('Disallowed character references MUST NOT be used. Only hexadecimal numeric character references '
+                  '(e.g. "&#x3A3;") and certain named references ("&lt;", "&gt;", "&amp;", "&apos;", and "&quot;") are allowed.'),
+            sourceFileLines=sourceFileLines
```

Comment:
This is fine, but for documents where the entire document is printed on one line this is not going to be very helpful because the source line will just be line 1, correct?

Reply:
I'll add `foundChars` like I have in 1.02